spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,292 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A mirco-batch based Spark Structured Streaming reader for Iceberg table. It will track the added
+ * files and generate tasks per batch to process newly added files. By default it will process
+ * all the newly added files to the current snapshot in each batch, user could also set this
+ * configuration "max-files-per-trigger" to control the number of files processed per batch.
+ */
+class StreamingReader extends Reader implements MicroBatchReader {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReader.class);
+  private static final int DEFAULT_MAX_FILES_PER_TRIGGER = 1000;
+  private static final long CACHE_SIZE = 1000;
+
+  private Offset startOffset;
+  private Offset endOffset;
+
+  private final Table table;
+  private final int maxFilesPerTrigger;
+  private final long splitSize;
+  private final int lookback;
+  private final long openFileCost;
+  private final Offset initialOffset;
+
+  private final Cache<Long, CloseableIterable<FileScanTask>> snapshotTasksCache = Caffeine.newBuilder()
+      .maximumSize(CACHE_SIZE)
+      .softValues()
+      .build();
+
+  /**
+   * Utility class to track the snapshotId, task and index of task within this snapshot.
+   */
+  @VisibleForTesting
+  static class IndexedTask {
+    private final long snapshotId;
+    private final FileScanTask task;
+    private final int index;
+
+    IndexedTask(long snapshotId, FileScanTask task, int index) {
+      this.snapshotId = snapshotId;
+      this.task = task;
+      this.index = index;
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    FileScanTask task() {
+      return task;
+    }
+
+    int index() {
+      return index;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxFilesPerTrigger =
+      options.get("max-files-per-trigger").map(Integer::parseInt).orElse(DEFAULT_MAX_FILES_PER_TRIGGER);
+    Preconditions.checkArgument(maxFilesPerTrigger > 0,
+        "Option max-files-per-trigger '%d' should > 0", maxFilesPerTrigger);
+
+    Long snapshotId = options.get("snapshot-id").map(Long::parseLong).orElse(null);
+    if (snapshotId != null) {
+      if (SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), snapshotId)) {
+        this.initialOffset = new StreamingOffset(snapshotId, -1, true);
+      } else {
+        LOG.warn("The option snapshot-id {} is not a valid snapshot id, will use the earliest " +
+            "snapshot instead", snapshotId);
+        this.initialOffset = calculateInitialOffset();
+      }
+    } else {
+      this.initialOffset = calculateInitialOffset();
+    }
+
+    this.splitSize = Optional.ofNullable(table.properties().get(TableProperties.SPLIT_SIZE))
```

Comment:
In the batch reader, `options` can override these values. Should that be supported here as well?